tst/unit/test_container_iterator.cpp

```diff
@@ -67,72 +82,383 @@ TEST_CASE("Can pull variables from containers based on Metadata", "[ContainerIte
-      }
 
-      THEN("they should sum appropriately") {
+        using policy4D = Kokkos::MDRangePolicy<Kokkos::Rank<4>>;
         Real total = 0.0;
-        for (int n = 0; n < civ.size(); n++) {
-          using policy4D = Kokkos::MDRangePolicy<Kokkos::Rank<4>>;
-          ParArrayND<Real> v = civ[n]->data;
-          Real sum = 0.0;
-          Kokkos::parallel_reduce(
-              policy4D({0, 0, 0, 0},
-                       {v.GetDim(4), v.GetDim(3), v.GetDim(2), v.GetDim(1)}),
-              KOKKOS_LAMBDA(const int l, const int k, const int j, const int i,
-                            Real &vsum) { vsum += v(l, k, j, i); },
-              sum);
-          total += sum;
-        }
-        REQUIRE(std::abs(total - 20480.0) < 1.e-14);
+        Real sum = 1.0;
+        Kokkos::parallel_reduce(
+            policy4D({0, 0, 0, 0},
+                     {v.GetDim(4), v.GetDim(3) - 1, v.GetDim(2) - 1, v.GetDim(1) - 1}),
+            KOKKOS_LAMBDA(const int l, const int k, const int j, const int i,
+                          Real &vsum) { vsum += v(l, k, j, i); },
+            sum);
+        total += sum;
+        REQUIRE(std::abs(total - 50625.0) < 1.e-14);
       }
     }
+
+    WHEN("we add sparse fields") {
+      Metadata m_sparse;
+      m_sparse = Metadata({Metadata::Sparse}, 1);
+      rc.Add("vsparse", m_sparse, scalar_block_size);
+      m_sparse = Metadata({Metadata::Sparse}, 13);
+      rc.Add("vsparse", m_sparse, scalar_block_size);
+      m_sparse = Metadata({Metadata::Sparse}, 42);
+      rc.Add("vsparse", m_sparse, scalar_block_size);
+      THEN("the low and high index bounds are correct as returned by PackVariables") {
+        PackIndexMap imap;
+        auto v = PackVariables<Real>(rc, {"v3", "v6", "vsparse"}, imap);
+        REQUIRE(imap["vsparse"].first == 4);
+        REQUIRE(imap["vsparse"].second == 6);
+      }
+      AND_THEN("bounds are still correct if I get just a subset of the sparse fields") {
+        PackIndexMap imap;
+        auto v = PackVariables<>(rc, {"v3", "vsparse"}, {1, 42}, imap);
+        REQUIRE(imap["vsparse"].first == 3);
+        REQUIRE(imap["vsparse"].second == 4);
+        REQUIRE(imap["vsparse_42"].first == 4);
+      }
+    }
+  }
+}
+// Test wrapper to run a function multiple times
+template <typename InitFunc, typename PerfFunc>
+double performance_test_wrapper(const int n_burn, const int n_perf, InitFunc init_func,
+                                PerfFunc perf_func) {
+  // Initialize the timer and test
+  Kokkos::Timer timer;
+  init_func();
+
+  for (int i_run = 0; i_run < n_burn; i_run++) {
+    // burn
+    perf_func();
+  }
+  Kokkos::fence();
+  timer.reset();
+  for (int i_run = 0; i_run < n_perf; i_run++) {
+    // time
+    perf_func();
+  }
+
+  // Time it
+  Kokkos::fence();
+  double perf_time = timer.seconds();
+
+  // FIXME?
+  // Validate results?
+
+  return perf_time;
+}
+
+TEST_CASE("Container Iterator Performance", "[ContainerIterator][performance]") {
```

Comment:
Should we incorporate some of @forrestglines separate performance tests directly into the unit tests here?

Reply:
I think we have enough testing here to catch an accidental significant change in performance. More rigorous performance testing might add too much overhead for the test suite.